README.md

```diff
@@ -1,78 +1,113 @@
 
-* we use `isequal()` because we want to match `Inf` values, etc.
+## User-specified hash function
 
-* we include the type in the hash so that different types with the same
-  contents don't collide
+You can specify the hash function to be implemented, by naming it before the struct definition with a keyword argument `hashfn`:
 
-* the type and `true` make it simple to generate code for empty records
+```julia
+@auto_hash_equals hashfn=SomePackage.myhash struct Foo
+    x
+    y
+end
+```
 
-* the `Base` module is explicitly used so that you don't need to
-  import it
+In this case the macro implements both `SomePackage.myhash` and `Base.hash` for `Foo`.`
 
-## Background
+## Caching the hash value
 
-Julia has two composite types: *value* types, defined with `struct`, and
-*record* types, defined with `mutable struct`.
+You can have the hash value precomputed and stored in a hidden field, by adding the keyword argument `cache=true`. This useful for non-mutable struct types that define recursive or deep data structures (and therefore are likely to be stored on the heap).  It computes the hash code during construction and caches it in a field of the struct.  If you are working with data structures of any significant depth, computing the hash once can speed things up at the expense of one additional field per struct.
 
-Value types are intended for compact, immutable objects.  They are stored on
-the stack, passed by value, and the default hash and equality are based on the
-literal bits in memory.
+```julia
+@auto_hash_equals cache=true struct Box{T}
+    x::T
+end
+```
 
-Record types are allocated on the heap, are passed by reference, and the
-default hash and equality are based on the pointer value (the data address).
+this translates to
 
-When you embed a record type in a value type, then the pointer to the record
-type becomes part of the value type, and so is included in equality and hash.
+```julia
+struct Box{T}
+    x::T
+    _cached_hash::UInt
+    function Box{T}(x) where T
+        new(x, Base.hash(x, Base.hash(Box{T})))
+    end
+end
+function Base.hash(x::Box, h::UInt)
+    Base.hash(x._cached_hash, h)
+end
+function Base.hash(x::Box)
+    x._cached_hash
+end
+function Base._show_default(io::IO, x::Box)
+    AutoHashEqualsCached._show_default_auto_hash_equals_cached(io, x)
+end
+function Base.:(==)(a::Box{T}, b::Box{T}) where T
+    a._cached_hash == b._cached_hash && Base.isequal(a.x, b.x)
+end
+function Box(x::T) where T
+    Box{T}(x)
+end
+```
 
-Given the above, it is often necessary to define hash and equality for
-composite types.  Particularly when record types are used (directly, or in a
-value type), and when records with the same contents are semantically equal.
+When you use the cache mechanism, the generated code checks the exact types of the two compared objects, so `Box{Int}(1)` will compare unequal to `Box{Any}(1)`.
```

Comment:
Perhaps this difference between the cache and non-cache case is now gratuitous.  Should I always make it take the type arguments into account, or never, or make it another option you can control?

Reply:
what was the original thinking on why uses types and the other doesn't? It does feel a bit off that they mismatch here and not via a user option

Reply:
Complex has complex semantics (no pin intended).  Here's another data point:

```julia
julia> struct Box{T}; x::T; end

julia> isequal(Box{Any}(1), Box{Int}(1))
false

julia> hash(Box{Any}(1))
0x552d7de52ee72c93

julia> hash(Box{Int}(1))
0x9f55fa74be71de5b
```

Reply:
of course the example I pull out of thin air as it's a parameterized type in base has weird special casing 🤦 

If `isequal` takes into account the types, that does make me lean toward taking them into account in hash equality as well. Though I need to think a bit more about the semantics of how that plays with the fields-we-care-about-majigger and if that influences the decision the other way 🤔

Reply:
OK, adding a keyword option 
> * `typearg=true|false` whether or not to make type arguments significant. Default: `true`.

Reply:
this comment I think is now out of date?